Bugfix/ TLVF: empty class member initialization

CoralMalachi opened the pull request:
When a TLV contains a class member (lets call it var1), since it might
contain a variable-length/unknown-length lists, its its init() function
doesn't create var1 when it is first created in a CmduMessageTx (m_parse == false).

However, when the same TLV is created in a CmduMessageRx (m_parse == true),
var1 is created automatically by the init() function since its length is
already known.

Due to this, when creating a CmduMessageTx containing a class member, the
caller needs to create-add it even if it is not used.

In the next commits, Finalize will fail with a clear
error message if any TLV inside the CMDU is not fully initialized.

In order to illustrate the situation of an error case, remove the
manually adding of var1, and change the finalize() to fail if not each of the members of the type class in TLV is fully initialized

Fixes #480

Commit:
tlvf: test: remove manually adding of var1

When a TLV contains a class member (lets call it var1), since it might
contain a variable-length/unknown-length lists, its init() function
doesn't create var1 when it is first created in a CmduMessageTx (m_parse == false).

However, when the same TLV is created in a CmduMessageRx (m_parse == true),
var1 is created automatically by the init() function since its length is
already known.

Due to this, when creating a CmduMessageTx containing a class member, the
caller needs to create-add it even if it is not used.

In the next commits, Finalize will fail with a clear
error message if any TLV inside the CMDU is not fully initialized.

In order to illustrate the situation of an error case, remove the
manually adding of var1.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: BaseClass: add is post init succeeded virtual method

Preparative commit.
In order to validate in the Finalize action if there is any class member
that is not fully initialized, a bool member is added to the BaseClass,
and a method to return its value.

The method is virtual since it is going to be overridden in case of a TLV
contains a class member.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: override post init succeeded method

In order to make sure that a class that contains a class member is fully
initialized, the virtual method isPostInitSucceeded is overridden to
validate that and return true or false to indicate.

For that reason, the following changes were made:

For each TLV that contains a member of a type class, a private bool member
is created (for each class member) which indicates if that member class is
initialized. This bool variable is initialized to FALSE.
Also, a method to return the value of that member is added.

When the user calls the add_member method, the value of the bool is set
to TRUE.

In those cases, the method isPostInitSucceeded is overridden to include
more conditions to validate that each of the members of the type class
in TLV is fully initialized.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: CmduTx: add a validation check for member class in tlv

Since it is very hard to find the reason for the error when not adding a
member of type class manually, the Finalize method is changed to return
FALSE if any TLV contains a member of type class which is not fully
initialized, and print a clear error message.

Now, when running the `tlvf_test` without add var1 manually, a clear
error message is printed to the screen.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
framework: tlvf_test: readd the manually adding of var1

Readd the manually adding of var1.
Now, when running the `tlvf_test` there is no error as expected.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

morantr commented on framework/tlvf/AutoGenerated/src/BaseClass.cpp:
```diff
@@ -13,7 +13,7 @@
 BaseClass::BaseClass(uint8_t *buff, const size_t buff_len, const bool parse, const bool swap_needed)
     : m_buff__(buff), m_buff_ptr__(buff), m_buff_len__(buff_len), m_parse__(parse),
-      m_swap__(swap_needed), m_init_succeeded(false)
+      m_swap__(swap_needed), m_init_succeeded(false) , m_post_init_succeeded(true)
```
There is no point in initializing it here to a default value. It could be done in the member declaration. Same for the `m_init_succeeded`.
(nitpick)

morantr commented on framework/tlvf/src/include/tlvf/BaseClass.h:
```diff
@@ -28,6 +28,7 @@ class BaseClass {
 
     size_t getLen();
     bool isInitialized();
+    virtual bool isPostInitSucceeded() { return m_post_init_succeeded; };
```
If you want to override it, then why did you do an implementation here? 
I think that the right thing is to define it as pure virtual with no implementation.

CoralMalachi replied:
Since I want to override this method only in cases that the TLV contains a member of type class.

morantr commented on framework/tlvf/AutoGenerated/src/tlvf/test/tlvVarList.cpp:
```diff
@@ -225,6 +225,13 @@ bool tlvTestVarList::add_complex_list(std::shared_ptr<cInner> ptr) {
 }
 
+bool tlvTestVarList::isPostInitSucceeded() {
+    if( !is_var1_initialized() ){
```
Please remove the spaces between the `if` brackets to the internal condition, and add space between the `if` to the open bracket.

morantr commented on framework/tlvf/AutoGenerated/src/tlvf/test/tlvVarList.cpp:
```diff
@@ -225,6 +225,13 @@ bool tlvTestVarList::add_complex_list(std::shared_ptr<cInner> ptr) {
 
+bool tlvTestVarList::isPostInitSucceeded() {
+    if( !is_var1_initialized() ){
+        m_post_init_succeeded=false;
```
Please add space before and after the `=` sign.

morantr commented on framework/tlvf/AutoGenerated/src/tlvf/test/tlvVarList.cpp:
```diff
@@ -266,6 +273,7 @@ bool tlvTestVarList::add_var1(std::shared_ptr<cInner> ptr) {
         TLVF_LOG(ERROR) << "Not enough available space on buffer";
         return false;
     }
+    m_var1_init = true;
```
I gotta say I don't fully understand what is the difference between setting this boolean to true, to return true in the function.
If the function is failing, then this member will also be set to false.

tomereli replied:
When you return true, the caller knows that add succeeded. But if the caller never called the add function, tlvf needs to be aware of that so in Finalize we check the value of this variable.

tomereli reviewed (commented):
I would change the start of the first commit message (and PR description):

```
When a TLV contains a class member (lets call it var1), since it might contain a
variable-length/unknown-length lists, its its init() function
doesn't create var1 when it is first created in a CmduMessageTx (m_parse == false).

However, when the same TLV is created in a CmduMessageRx (m_parse == true), var1 is created
automatically by the init() function since its length is already known.

Due to this, when creating a CmduMessageTx containing a class member, the caller needs to
create-add it even if it is not used.
```

The rest of the

Commit:
tlvf: demonstrate a TLV with two class members

Add to tlvVarList another member of a type class, to demonstrate a case
of TLV that contains more than one member of type class.

The tlvf_test has been changed accordingly.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

tomereli reviewed (commented):
- commit e8afd796d57506b85e4ae10e147fd33deea7a2db: `isPostInitSucceede` should be `isPostInitSucceeded`

- each of the members of the type class in TLV **are** fully initialized.

tomereli commented on framework/tlvf/AutoGenerated/include/tlvf/ieee_1905_1/tlvWscM1.h:
```diff
@@ -88,6 +88,8 @@ class tlvWscM1 : public BaseClass
         WSC::sWscAttrDevicePasswordID& device_password_id_attr();
         WSC::sWscAttrConfigurationError& configuration_error_attr();
         WSC::sWscAttrOsVersion& os_version_attr();
+        bool is_vendor_ext_initialized() { return m_vendor_ext_init;}
```
No need for a function, you can access these member variables from `isPostInitSucceeded()` in the derived class.

tomereli commented on framework/tlvf/AutoGenerated/src/tlvf/ieee_1905_1/tlvWscM1.cpp:
```diff
@@ -465,6 +465,13 @@ WSC::sWscAttrOsVersion& tlvWscM1::os_version_attr() {
 }
 
+bool tlvWscM1::isPostInitSucceeded() {
+    if( !is_vendor_ext_initialized() ){
```
Just check `m_vendor_ext_init` instead of using a function here.

tomereli replied:
Remove the space before the `!`. Remove the space before the `)`. Add a space before the `{`
So:
```
if (!is_vendor_ext_initialized()) {
```

(obviously should be done in tlvf.py)

tomereli commented on framework/tlvf/AutoGenerated/src/tlvf/ieee_1905_1/tlvWscM1.cpp:
```diff
@@ -465,6 +465,13 @@ WSC::sWscAttrOsVersion& tlvWscM1::os_version_attr() {
 
+bool tlvWscM1::isPostInitSucceeded() {
+    if( !is_vendor_ext_initialized() ){
+        m_post_init_succeeded=false;
```
add a space before and after the `=`

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -209,6 +209,7 @@ def __init__(self, fname, name, dict):
         self.list_index = 0
         self.lock_allocation_member_added = False
         self.lock_order_member_added = False
+        self.marker_post_init_method_added = False
```
Can we call it `post_init_method_added` instead and set it to True after we encounter the first class member and really add the marker and the function?

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -336,7 +337,7 @@ def __init__(self, src_path, yaml_path, out_path, conf_path, print_dependencies,
         self.CODE_CLASS_PRIVATE_FUNC_INSERT         = "//~class_private_func_insert"
         self.CODE_CLASS_INIT_FUNC_INSERT            = "//~class_init_func_insert"
-        self.CODE_CLASS_INIT_FUNC_SWAP_INSERT            = "//~class_init_func_swap_insert"
+        self.CODE_CLASS_INIT_FUNC_SWAP_INSERT       = "//~class_init_func_swap_insert"
```
Should be on a separate commit (nitpick)

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -346,6 +347,7 @@ def __init__(self, src_path, yaml_path, out_path, conf_path, print_dependencies,
         self.CODE_STRUCT_REVERSED_BITFIELD_INSERT   = "//~struct_reversed_bitfield_insert"
         self.CODE_STRUCT_SWAP_FUNC_INSERT           = "//~struct_swap_func_insert"
         self.CODE_STRUCT_INIT_FUNC_INSERT           = "//~struct_init_func_insert"
+        self.CODE_CLASS_IS_INIT_FUNC                = "//~class_is_init_func"
```
Please move this to after line 342 so it will be with its family :)

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -644,6 +646,41 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
                 lines_h.append("std::shared_ptr<%s> m_%s_ptr = nullptr;" % (param_type, param_name))
+                lines_h.append("bool m_%s_init = false;" % (param_name))
+                
+                #Add a get method in the H file
```
Should go into a function with documentation explaining all we are doing here and why. Its about time that tlvf gets some good comments.

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -644,6 +646,41 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
             if TypeInfo(param_type).type == TypeInfo.CLASS:                
                 lines_h.append("%s *m_%s = nullptr;" % (param_type, param_name))
                 lines_h.append("std::shared_ptr<%s> m_%s_ptr = nullptr;" % (param_type, param_name))
+                lines_h.append("bool m_%s_init = false;" % (param_name))
```
This should go to the same function below

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -1057,6 +1094,8 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
                 if is_var_len:
                     lines_cpp.append( "%sif (!m_parse__) { (*m_%s)++; }" % (self.getIndentation(1), param_length) )
+            else: 
+                lines_cpp.append("%sm_%s_init = true;" %(self.getIndentation(1),param_meta.name))
```
please add a comment here which explains this - refer to the function which adds the member that you will add later.

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -1167,7 +1206,6 @@ def addAllocationMarkersAlloc(self, obj_meta, param_meta, param_length, memmove)
             marker = "%s_%s_%s_%s" %(self.CODE_CLASS_ALLOC_INSERT, obj_meta.name, "alloc", param_meta.name)
             obj_meta.alloc_list.append((marker, "alloc"))
             lines_cpp.append(marker)
-
```
why remove the line break? same goes to the below.

tomereli commented on framework/tlvf/AutoGenerated/src/CmduMessageTx.cpp:
```diff
@@ -88,6 +88,10 @@ bool CmduMessageTx::finalize(bool swap_needed)
 
     // Update vendor specific length which is seperated to 3 classes
     for (auto class_it = m_class_vector.begin(); class_it != m_class_vector.end(); class_it++) {
+        if(!(*class_it)->isPostInitSucceeded()){
```
Is there any way we can add a print here saying which class postInit failed? Otherwise the error print will not say much.
Perhaps @arnout has an idea..

(nitpick)

tomereli commented on framework/tlvf/test/tlvf_test.cpp:
```diff
@@ -303,7 +303,10 @@ int test_parser()
 
     LOG(DEBUG) << "Finalize";
-    msg.finalize(true);
+    if (!msg.finalize(true)) {
```
Just to make sure - do we have this in the controller and agent?

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -583,6 +583,62 @@ def addStructParam(self, obj_meta, param_name, param_type, param_type_info, para
         else:
             self.insertLineH(obj_meta.name, self.CODE_STRUCT_INSERT, line)
 
+    ##########################################################################################
```
BIG LIKE!

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -583,6 +583,62 @@ def addStructParam(self, obj_meta, param_name, param_type, param_type_info, para
+            #Override isPostInitSucceeded method to verify the class is completly initialized
+            lines_cpp.append("bool %s::isPostInitSucceeded() {" % (obj_meta.name))
+            #if this member is not added by the user return false
+            lines_cpp.append("%sif(!m_%s_init) {" % (self.getIndentation(1) ,param_name))
```
we don't run clang-format here but we can at least try to align to our coding rules, so add a space after the `if`:
```
lines_cpp.append("%sif (!m_%s_init) {" % (self.getIndentation(1) ,param_name))
```

Same comment for other places you introduced an `if` without a space.

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -583,6 +583,62 @@ def addStructParam(self, obj_meta, param_name, param_type, param_type_info, para
+            #before the return statement
+            marker = "%s_%s" %(self.CODE_CLASS_IS_INIT_FUNC, obj_meta.name)
+            lines_cpp.append(marker)
+
```
why all these line breaks? (nitpick)

tomereli commented on framework/tlvf/tlvf.py:
```diff
@@ -1095,6 +1118,8 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
                     lines_cpp.append( "%sif (!m_parse__) { (*m_%s)++; }" % (self.getIndentation(1), param_length) )
             else: 
+                # When the user calls the add_member method, the value of the bool variable is set to TRUE.
+                # Fore more information, please read overrideIsPostInitSucceeded method documentation 
```
`Fore` --> `For`

tomereli commented on framework/tlvf/AutoGenerated/src/CmduMessageTx.cpp:
```diff
@@ -88,6 +88,10 @@ bool CmduMessageTx::finalize(bool swap_needed)
     // Update vendor specific length which is seperated to 3 classes
     for (auto class_it = m_class_vector.begin(); class_it != m_class_vector.end(); class_it++) {
+        if(!(*class_it)->isPostInitSucceeded()){
+            TLVF_LOG(ERROR) << "A TLV with a class member is not fully initialized";
```
Log should be more generic, saying "TLV post init failed".

tomereli reviewed (approved):
Mainly nitpicks and small typos changes, so please fix before merging, but otherwise LGTM. Nice
